azure-iot-device/azure/iot/device/iothub_session.py

```diff
@@ -84,47 +83,49 @@ def __init__(
         :keyword bool websockets: Set to 'True' to use WebSockets over MQTT. Default is 'False'
 
         :raises: ValueError if an invalid combination of parameters are provided
-        :raises: ValueError if an invalid 'symmetric_key' is provided
-        :raises: TypeError if an invalid keyword argument is provided
+        :raises: ValueError if an invalid parameter value is provided
+        :raises: TypeError if an unsupported keyword argument is provided
         """
         # Validate parameters
         _validate_kwargs(**kwargs)
-        if shared_access_key and sastoken_fn:
+        if shared_access_key and sastoken:
             raise ValueError(
-                "Incompatible authentication - cannot provide both 'shared_access_key' and 'sastoken_fn'"
+                "Incompatible authentication - cannot provide both 'shared_access_key' and 'sastoken'"
             )
-        if not shared_access_key and not sastoken_fn and not ssl_context:
+        if not shared_access_key and not sastoken and not ssl_context:
             raise ValueError(
-                "Missing authentication - must provide one of 'shared_access_key', 'sastoken_fn' or 'ssl_context'"
+                "Missing authentication - must provide one of 'shared_access_key', 'sastoken' or 'ssl_context'"
             )
 
-        # Set up SAS auth (if using)
-        generator: Optional[st.SasTokenGenerator]
-        # NOTE: Need to keep a reference to the SasTokenProvider so we can stop it during cleanup
-        self._sastoken_provider: Optional[st.SasTokenProvider]
+        if not ssl_context:
+            ssl_context = _default_ssl_context()
+
+        # Set up SAS auth for future use (if using)
+        self._user_sastoken: Optional[st.SasToken] = None
+        self._sastoken_generator: Optional[st.SasTokenGenerator] = None
+        self._sastoken_ttl = sastoken_ttl
         if shared_access_key:
+            # If using SasToken generation, we cannot generate during this __init__
+            # because .generate_sastoken() is a coroutine. The token will be generated and then
+            # set on the underlying client upon context manager entry
             uri = _format_sas_uri(hostname=hostname, device_id=device_id, module_id=module_id)
             signing_mechanism = sm.SymmetricKeySigningMechanism(shared_access_key)
-            generator = st.InternalSasTokenGenerator(
-                signing_mechanism=signing_mechanism, uri=uri, ttl=sastoken_ttl
+            self._sastoken_generator = st.SasTokenGenerator(
+                signing_mechanism=signing_mechanism, uri=uri
             )
```

Comment:
We still need sastoken_ttl, even if we don't have automatic renewal. The token still needs to expire eventually and we still want the customer to be able to se the expiration interval.

Reply:
It's added upon each generation